packages/redis/src/RedisStreamSink.ts

```diff
@@ -0,0 +1,96 @@
+import {
+    IOutputSink,
+    IPublishedMessage,
+    IOutputSinkGuarantees,
+    IRequireInitialization,
+    IDisposable,
+    IComponentContext,
+    makeLifecycle,
+    Lifecycle,
+    IMetrics,
+    failSpan,
+    DefaultComponentContext,
+    OutputSinkConsistencyLevel,
+} from "@walmartlabs/cookie-cutter-core";
+import { Span, Tags, Tracer } from "opentracing";
+
+import { IRedisInputStreamOptions, redisClient, IRedisClient, RedisMetadata } from ".";
+import { RedisOpenTracingTagKeys, RedisMetrics, RedisMetricResults } from "./RedisClient";
+
+export class RedisStreamSink
+    implements IOutputSink<IPublishedMessage>, IRequireInitialization, IDisposable {
+    public guarantees: IOutputSinkGuarantees;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private metrics: IMetrics;
+    private spanOperationName: string = "Redis Output Sink Client Call";
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+        this.metrics = DefaultComponentContext.metrics;
+        this.guarantees = {
+            consistency: OutputSinkConsistencyLevel.None,
+            idempotent: false,
+        };
+    }
+    async sink(
+        output: IterableIterator<IPublishedMessage>,
+        bail: (err: any) => never
+    ): Promise<void> {
+        let span: Span;
+        try {
+            for (const msg of output) {
+                span = this.tracer.startSpan(this.spanOperationName, {
+                    childOf: msg.spanContext,
+                });
+
+                this.spanLogAndSetTags(span, this.config.db, this.config.writeStream);
+
+                await this.client.xAddObject(
+                    span,
+                    msg.message.type,
+                    this.config.writeStream,
+                    RedisMetadata.OutputSinkStreamKey,
+                    msg.message.payload
+                );
+
+                this.metrics.increment(RedisMetrics.XAdd, {
+                    bucket: this.config.db,
+                    streamName: this.config.writeStream,
+                    result: RedisMetricResults.Success,
+                });
+
+                span.finish();
+            }
+        } catch (err) {
+            this.metrics.increment(RedisMetrics.XAdd, {
+                bucket: this.config.db,
+                streamName: this.config.writeStream,
+                result: RedisMetricResults.Error,
+            });
+
+            failSpan(span, err);
+            span.finish();
+            bail(err);
```

Comment:
are there any errors we know are retryable? we would want to just throw those vs calling bail (that way the retry logic can kick in).

also fyi: when you rebase this against `develop` the signature for sink slightly changed, and it's `retry.bail(err)` now (though your existing code would still work)